lib/core/buffer.hpp

```diff
@@ -41,30 +47,216 @@
 /// The system page size, this might become more dynamic someday
 #define PAGESIZE_MEM 4096
 
-/// The max length of a stage (consumer or producer) name.
-#define MAX_STAGE_NAME_LEN 128
-
-/// The maximum number of consumers that can register on a buffer
-#define MAX_CONSUMERS 20
-/// The maximum number of producers that can register on a buffer
-#define MAX_PRODUCERS 10
-
 /**
  * @struct StageInfo
  * @brief Internal structure for tracking consumer and producer names.
  */
-struct StageInfo {
-    /// Set to 1 if the stage is active
-    int in_use;
+class StageInfo {
+public:
+    StageInfo(const std::string& _name, int _num_frames) :
+        name(_name), num_frames(_num_frames), last_frame_acquired(-1), last_frame_released(-1),
+        is_done(_num_frames) {}
+    // No copies or moves allowed
+    StageInfo(StageInfo&&) = delete;
+    StageInfo& operator=(StageInfo&&) = delete;
+    StageInfo(const StageInfo&) = delete;
+    StageInfo& operator=(const StageInfo&) = delete;
 
     /// The name of the stage (consumer or producer)
-    char name[MAX_STAGE_NAME_LEN];
+    const std::string name;
+
+    /// The number of frames being processed by this object
+    int num_frames;
 
     /// Last frame acquired with a call to wait_for_*
     int last_frame_acquired;
 
     /// Last frame to be released with a call to mark_frame_*
     int last_frame_released;
+
+    // Is this producer/consumer done with each frame?
+    std::vector<bool> is_done;
+};
+
+class GenericBuffer : public kotekan::kotekanLogging {
+public:
+    GenericBuffer(const std::string&, const std::string&, metadataPool* pool, int num_frames);
+    virtual ~GenericBuffer();
+    virtual void print_full_status(){};
+
+    virtual bool is_basic() = 0;
```

Comment:
Yeah, the `is_basic()` is a hack to check whether a `GenericBuffer` instance is a (Frame) `Buffer`, for use in the few places that want data that isn't in `GenericBuffer`.

Ideally I would turn these things into generic functions that subclassers can implement as appropriate, but in practice those are a bit messy -- a good example is `stages/bufferStatus.cpp`, which is creating a `Metrics` object for the number of full frames in the buffers.  `RingBuffer` buffers don't have the concept of "full frames"

Reply:
I ended up removing `is_basic()` as a member variable, instead adding a free function `is_frame_buffer`, and also pulled out some of the Dot and JSON formatting into member functions so that RingBuffer can provide its own values (in the future; I didn't implement anything yet).